### PR TITLE
feat(identity): implement from_protobuf_encoding for RSA keys

### PR DESCRIPTION
**Pull request by mxinden**
## Description

Add support for reading a libp2p RSA keypair from a Protobuf.

## Notes & open questions

I need this for https://github.com/protocol/bifrost-infra/issues/2622. We want to replace an existing IPFS bootstrap node running [Kubo](https://github.com/ipfs/kubo) with [rust-libp2p-server](https://github.com/mxinden/rust-libp2p-server/). Due to legacy reasons IPFS bootstrap nodes use RSA. Thus [rust-libp2p-server](https://github.com/mxinden/rust-libp2p-server/) needs to be able to read an RSA key from disc.

See also corresponding specification: https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by mxinden** [comment]
@thomaseizinger follow-up commits here definitely showcase self-service CI in action :sparkles:

**Review by thomaseizinger — COMMENTED** [review]
Can you also update https://github.com/libp2p/specs/pull/537 and verify with other impls that this is a compliant encoding? I thought that ring does not support decoding our format.

**Comment by thomaseizinger** [comment]
Also, would this resolve https://github.com/libp2p/rust-libp2p/issues/3630?

**Comment by mxinden** [comment]
> I thought that ring does not support decoding our format.

I was thinking the same. We might be confusing decoding support with encoding support in ring?

**Comment by mxinden** [comment]
> Also, would this resolve #3630?

It does not. This pull request only implements the decoding not the encoding side.

File: identity/src/keypair.rs [hunk 1]
```diff
@@ -822,25 +830,24 @@ mod tests {
...
-        feature = "ed25519",
-        feature = "rsa",
-        feature = "peerid"
-    ))]
+    #[cfg(feature = "peerid")]
     fn keypair_from_protobuf_encoding() {
-        // E.g. retrieved from an IPFS config file.
-        let base_64_encoded = "CAESQL6vdKQuznQosTrW7FWI9At+XX7EBf0BnZLhb6w+N+XSQSdfInl6c7U4NuxXJlhKcRBlBw9d0tj2dfBIVf6mcPA=";
-        let expected_peer_id =
-            PeerId::from_str("12D3KooWEChVMMMzV8acJ53mJHrw1pQ27UAGkCxWXLJutbeUMvVu").unwrap();
```
Comment (mxinden) [inline identity/src/keypair.rs:836]: This pull request removes the `keypair_from_protobuf_encoding` `ed25519` test. In my eyes this is sufficiently covered with the more elaborate round-trip test introduced with https://github.com/libp2p/rust-libp2p/pull/4228.

File: identity/src/keypair.rs [hunk 2]
```diff
@@ -822,25 +830,18 @@ mod tests {
...
+            "080012a60430820222300d06092a864886f70d01010105000382020f003082020a0282020100e1beab071d08200bde24eef00d049449b07770ff9910257b2d7d5dda242ce8f0e2f12e1af4b32d9efd2c090f66b0f29986dbb645dae9880089704a94e5066d594162ae6ee8892e6ec70701db0a6c445c04778eb3de1293aa1a23c3825b85c6620a2bc3f82f9b0c309bc0ab3aeb1873282bebd3da03c33e76c21e9beb172fd44c9e43be32e2c99827033cf8d0f0c606f4579326c930eb4e854395ad941256542c793902185153c474bed109d6ff5141ebf9cd256cf58893a37f83729f97e7cb435ec679d2e33901d27bb35aa0d7e20561da08885ef0abbf8e2fb48d6a5487047a9ecb1ad41fa7ed84f6e3e8ecd5d98b3982d2a901b4454991766da295ab78822add5612a2df83bcee814cf50973e80d7ef38111b1bd87da2ae92438a2c8cbcc70b31ee319939a3b9c761dbc13b5c086d6b64bf7ae7dacc14622375d92a8ff9af7eb962162bbddebf90acb32adb5e4e4029f1c96019949ecfbfeffd7ac1e3fbcc6b6168c34be3d5a2e5999fcbb39bba7adbca78eab09b9bc39f7fa4b93411f4cc175e70c0a083e96bfaefb04a9580b4753c1738a6a760ae1afd851a1a4bdad231cf56e9284d832483df215a46c1c21bdf0c6cfe951c18f1ee4078c79c13d63edb6e14feaeffabc90ad317e4875fe648101b0864097e998f0ca3025ef9638cd2b0caecd3770ab54a1d9c6ca959b0f5dcbc90caeefc4135baca6fd475224269bbe1b0203010001"
+        ))
+            .unwrap();
```
Comment (mxinden) [inline identity/src/keypair.rs:842]: The pull request then adds a non-round-trip test for `rsa` using the fixtures from https://github.com/libp2p/specs/pull/537#pullrequestreview-1539424662.

Note that this is not doing a round-trip test as we don't currently implement the encoding of `rsa` keys to our protobuf encoding, from only implement from the protobuf encoding.

**Review by mxinden — COMMENTED** [review]
I forgot about https://github.com/libp2p/specs/pull/537. Good idea @thomaseizinger.

Would you mind taking another look?

**Comment by thomaseizinger** [comment]
> > I thought that ring does not support decoding our format.
> 
> I was thinking the same. We might be confusing decoding support with encoding support in ring?

Ah yes, exactly. And we didn't want to convert between the two formats.

**Review by thomaseizinger — APPROVED** [review]
Thanks Max! I think we should merge #4228 first.

**Pull request merged**